Atangster/132 add callback functions to grantpermissionui and appbar

atangster opened the pull request:
## Pull Request Details

Link to associated issue: https://github.com/anusii/moviestar/issues/132#issuecomment-3160350800

# Checklist
Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [x] Screenshots included in linked issue
- [x] Changes adhere to the team style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [x] No lint check errors related to your changes (make prep or flutter analyze lib)
- [x] Tested on at least one device
  - [x] Windows
  - [x] Linux
- [x] Added 2 reviewers (or 1 for private repositories then they add another)

atangster committed:
add callback functions to grantpermissionui and appbar

atangster committed:
format code

atangster committed:
update comment

atangster committed:
Merge remote-tracking branch 'origin/dev' into atangster/132_update_ui

anushkavidanage commented on lib/src/solid/grant_permission_ui.dart:
```diff
@@ -102,6 +104,14 @@ class GrantPermissionUi extends StatefulWidget {
 
+  /// Callback function called when permissions are granted successfully.
+  
+  final VoidCallback? onPermissionGranted;
```
But where is this function called? I cannot seem to find a place in the code where this is called.

gjwgit replied:
More discussion/clarification required.

atangster replied:
From https://github.com/anusii/moviestar/pull/145

1. `lib\screens\to_watch_screen.dart`

```dart
onPermissionGranted: () async {
                  // After movie list sharing, chain individual movie sharing
                  if (mounted) {
                    Navigator.of(navContext).pop(true);

                    // Show loading and start individual movie sharing
                    await _shareIndividualMoviesSequentially(movies);
                  }
                },
                onNavigateBack: () {
                  // Handle back navigation
                  if (mounted) {
                    Navigator.of(navContext).pop(false);
                  }
```

2. `lib\screens\watched_screen.dart`

```dart
onPermissionGranted: () async {
                  // After movie list sharing, chain individual movie sharing
                  if (mounted) {
                    Navigator.of(navContext).pop(true);

                    // Show loading and start individual movie sharing
                    await _shareIndividualMoviesSequentially(movies);
                  }
                },
                onNavigateBack: () {
                  // Handle back navigation
                  if (mounted) {
                    Navigator.of(navContext).pop(false);
                  }
                },
```

atangster committed:
add example permission callback demo

atangster commented:
Thanks @anushkavidanage for the feedback

I've added a demo in `example` app for the permission callback

These points of discussion may also be of interest:
- Permission callback analysis https://github.com/anusii/moviestar/issues/132#issuecomment-3177934445
- Alternatives https://github.com/anusii/moviestar/issues/132#issuecomment-3177939282
- Future considerations https://github.com/anusii/moviestar/issues/132#issuecomment-3177946664

- [x] Permission callback is accessible via the example app home page

<img width="1245" height="1015" alt="Screenshot 2025-08-12 161459" src="https://github.com/user-attachments/assets/9a85033f-2c3d-4d70-ba20-7b729a87b36b" />

- [x] When user presses 'Permission Callback Demo` button, it takes them to the specific page with the demo
- [x] User presses Start Auto-Demo to begin

<img width="1274" height="1025" alt="Screenshot 2025-08-12 161516" src="https://github.com/user-attachments/assets/838a9c0a-c46e-4b7e-acab-2cb3842082c0" />

- [x] Demo files are created automatically during demo (assuming user has already entered security key)
- [x] User is prompted with individual grant permission UI for sharing the demo file
- [x] User to add a WebID on the server and READ permissions 
<img width="1264" height="1003" alt="Screenshot 2025-08-12 161533" src="https://github.com/user-attachments/assets/76cbff45-dae9-4f2e-b468-f24071218a8c" />

- [x] User goes through 3 continuous iterations of sharing, to share each of the three demo files
- [x] Once fully successful, user will see success snackbar and updated Permission Callback Demo page 

<img width="1250" height="1029" alt="Screenshot 2025-08-12 161617" src="https://github.com/user-attachments/assets/1e935ab1-219b-48d6-a6ed-e0ad0f799ca5" />

anushkavidanage commented:
Thanks @atangster, I tested your demo page and it worked great. Then I tried to add my own callback function to the `Add/Delete Permissions from any Resource` button and that unfortunately did not work. Is there something that I should do to add my own function? Here's the extra code I added.

```dart
onPermissionGranted: () {
    // Navigate back to the home screen.

    print(
        'Permission granted. Going back to previous screen');

    Navigator.of(context).pop(true);
  }
```

atangster commented:
> Is there something that I should do to add my own function?

@anushkavidanage Yes, the function signature is missing the `async` keyword, which is required since you're using `Navigator.of(context).pop(true)`.

Navigator operations should be awaited:
```dart
  onPermissionGranted: () async {
      print('Permission granted. Going back to previous
  screen');
      Navigator.of(context).pop(true);
  }
```

gjwgit commented:
No longer used in moviestar but a useful addition.
